src/image/detail/MaskDict.cc

```diff
@@ -31,18 +32,48 @@ class MaskDict::GlobalState {
 
     using HandleList = std::set<std::weak_ptr<MaskDict>, std::owner_less<std::weak_ptr<MaskDict>>>;
 
-    static GlobalState & get() {
-        static GlobalState instance;
-        return instance;
-    }
+    /*
+     * RAII guard for GlobalState.
+     *
+     * Default-constructing a Lock instance is the only way to access
+     * GlobalState - this acquires GlobalState's mutex, ensuring only one
+     * thread has access to its attributes at one time.  The mutex is released
+     * when the Lock instance goes out of scope.
+     */
+    class Lock {
+    public:
+
+        Lock() :
+            allMaskDicts(_get()._allMaskDicts),
+            defaultMaskDict(_get()._defaultMaskDict),
+            _guard(_get()._mutex)
+        {}
+
+        ~Lock() = default;
+
+        HandleList & allMaskDicts;
+        std::shared_ptr<MaskDict> & defaultMaskDict;
```

Comment:
I think I see... you mean that, even without the synchronization question, you want `defaultMaskDict` to be updated any time `_get()._defaultMaskDict` is? That's also worth documenting (basically, references to `shared_ptr` are weird, so an explanation would be helpful).

Reply:
Sorry, what I meant was that I wanted to allow resetting `_get()._defaultMaskDict` and modifying `_get()._allMaskDicts` within the scope of a single `Lock` instance (for `detachDefault`), so no other thread could see a change to GlobalState data member without the corresponding change to the other.  But now I'm thinking we are imagining totally different alternatives to what the code looks like now, and talking about different aspects of it.

Reply:
That's guaranteed as long as all access (including any reads) is guarded with the same lock (which it is, except for the reads in the `Lock` constructor).

But I just realized I was an idiot the first time around: `defaultMaskDict` must be a reference, because otherwise assignments to it won't affect the original pointer. 🤦‍♂️